### PR TITLE
Log - StdoutErrConsumer [9069]

### PR DESCRIPTION
**Pull request by EduPonz**
This PR adds a new Log consumer `StdoutErrConsumer` with the following features:
1. It outputs log messages to different streams depending on their severity. Log messages equal to or more severe than the threshold are output to `STDERR` using `std::cerr`. Log messages less severe than the threshold are output to `STDOUT` using `std::cout`.
1. The threshold can be set and retrieved using `StdoutErrConsumer::stderr_threshold()`.
1. Exposes a constant `STDERR_THRESHOLD_DEFAULT` that marks the default threshold value, which is `Log::Kind::Warning`.
1. `StdoutErrConsumer` consumers can also be registered using XML. In this case, the threshold is set using the property `stderr_threshold`.

The PR also:
1. Sets `StdoutErrConsumer` as the default log consumer.
1. Includes tests for `StdoutErrConsumer`.
1. Includes tests for registering `StdoutErrConsumer` using XML.
1. Better handling and propagation of different errors on parsing XML files.

Signed-off-by: EduPonz <eduardoponz@eprosima.com>

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2490/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2304/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2497/)

**Review by LuisGP — CHANGES_REQUESTED** [review]

File: test/unittest/rtps/writer/CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2016 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+# Copyright 2016, 2017, 2018, 2019, 2020 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply test/unittest/rtps/writer/CMakeLists.txt]: It seems to be enough to keep the year of first publication. It has been fixed in all the files.

File: test/unittest/xmlparser/log_stdouterr_wrong_property_name.xml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<dds>
+    <log>
+        <use_default>FALSE</use_default>
+        <consumer>
+            <class>StdoutErrConsumer</class>
+            <property>
+                <name>threshold</name>
```
Comment (MiguelCompany) [inline test/unittest/xmlparser/log_stdouterr_wrong_property_name.xml:8]:
```suggestion
                <name>wrong_property_name</name>
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2962/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/225/)
* Mac Build Status
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2974/)

File: src/cpp/fastdds/log/Log.cpp [hunk 3]
```diff
@@ -16,13 +32,13 @@ struct Log::Resources Log::resources_;
...
-    ,functions(true)
-    ,verbosity(Log::Error)
+    , work(false)
+    , current_loop(0)
+    , filenames(false)
+    , functions(true)
+    , verbosity(Log::Error)
 {
-    resources_.consumers.emplace_back(new StdoutConsumer);
+    resources_.consumers.emplace_back(new StdoutErrConsumer);
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply src/cpp/fastdds/log/Log.cpp:41]: A new CMake option is going to be implemented to allow the user to choose the default Log Consumer without having to change his application code. By default, this option will use the current default Log Consumer (StdoutConsumer) ensuring backwards compatibility.

File: src/cpp/rtps/xmlparser/XMLParser.cpp [hunk 4]
```diff
@@ -131,11 +132,19 @@ XMLP_ret XMLParser::parseXML(
                 }
                 else if (strcmp(tag, TYPES) == 0)
                 {
-                    ret = parseXMLTypes(node);
+                    up_base_node_t types_node = up_base_node_t{ new BaseNode{NodeType::TYPES} };
+                    if (XMLP_ret::XML_OK == (ret = parseXMLTypes(node)))
+                    {
+                        root->addChild(std::move(types_node));
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply src/cpp/rtps/xmlparser/XMLParser.cpp:141]: Comments have been added to the code addressing this questions. It is a workaround to propagate the XML return codes upstream. Only the child type is checked in `XMLProfileManager::loadXMLFile(const std::string& filename)` so the object does not need to be populated.

File: src/cpp/rtps/xmlparser/XMLParser.cpp [hunk 5]
```diff
@@ -131,11 +132,19 @@ XMLP_ret XMLParser::parseXML(
...
+                    }
                 }
                 else if (strcmp(tag, LOG) == 0)
                 {
-                    ret = parseLogConfig(node);
+                    up_base_node_t log_node = up_base_node_t{ new BaseNode{NodeType::LOG} };
+                    if (XMLP_ret::XML_OK == (ret = parseLogConfig(node)))
+                    {
+                        root->addChild(std::move(log_node));
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply src/cpp/rtps/xmlparser/XMLParser.cpp:152]: See answer above.

File: CMakeLists.txt [hunk 6]
```diff
@@ -266,6 +266,21 @@ if(EPROSIMA_BUILD AND NOT EPROSIMA_INSTALLER)
     set(SHM_TRANSPORT_DEFAULT ON)
 endif()
 
+###############################################################################
+# LogConsumer default setup
+###############################################################################
+set(STDOUT_LOG_CONSUMER OFF CACHE BOOL "Stdout LogConsumer as default" FORCE)
+set(STDOUTERR_LOG_CONSUMER OFF CACHE BOOL "StdoutErr LogConsumer as default" FORCE)
```
Comment (MiguelCompany) [inline CMakeLists.txt:273]: I don't think we need these to be on CACHE. Could be just normal variables.

```suggestion
```

File: CMakeLists.txt [hunk 7]
```diff
@@ -266,6 +266,21 @@ if(EPROSIMA_BUILD AND NOT EPROSIMA_INSTALLER)
...
+set(LOG_CONSUMER_DEFAULT AUTO CACHE STRING "Selects default LogConsumer")
+
+set_property(CACHE LOG_CONSUMER_DEFAULT PROPERTY STRINGS AUTO STDOUT STDOUTERR)
+
+if (LOG_CONSUMER_DEFAULT STREQUAL "AUTO" OR LOG_CONSUMER_DEFAULT STREQUAL "STDOUT")
+    set(STDOUT_LOG_CONSUMER ON CACHE BOOL "Stdout LogConsumer as default" FORCE)
+elseif(LOG_CONSUMER_DEFAULT STREQUAL "STDOUTERR")
+    set(STDOUTERR_LOG_CONSUMER ON CACHE BOOL "StdoutErr LogConsumer as default" FORCE)
+endif()
```
Comment (MiguelCompany) [inline CMakeLists.txt:282]: In fact, only one of them is necessary

```suggestion
if(LOG_CONSUMER_DEFAULT STREQUAL "STDOUTERR")
    set(STDOUTERR_LOG_CONSUMER ON)
endif()
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2996/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/243/)
* Mac Build Status
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3007/)

**Review by MiguelCompany — APPROVED** [review]

**Comment by JLBuenoLopez-eProsima** [comment]
Documentation pull request: eProsima /Fast-RTPS-docs#201

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3001/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/246/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2691/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3012/)

**Pull request merged**